### ISSUE TITLE
Image in raster layer shows only partially on certain zoom levels when 3D terrain is enabled.

### ISSUE THREAD

**Issue by danielfaust** [maplibre/maplibre-gl-js#1559]
![CEA884BC-D1C4-49DB-9611-8BBC39B1C392](https://user-images.githubusercontent.com/320098/187073663-e3bf2755-52ed-4056-a651-3352a907d728.GIF)

**maplibre-gl-js version**: v2.4.0 (already present in v2.2.0-pre.2)

**browser**: Google Chrome Version 104.0.5112.102 (Offizieller Build) (64-Bit)

No errors or warnings appear in the console

### Steps to Trigger Behavior

 1. Zoom in and out into the image to see the problem, then zoom in to prepare to toggle elevation layer
 2. Toggle elevation layer a couple of times to see the problem.
 
### Link to Demonstration

~https://jsbin.com/qeverim/edit?html,output~

https://jsbin.com/yowixah/1/edit?html,output

### Expected Behavior

The image should always fully render into the layer

### Actual Behavior

The image is cropped at high zooming levels.

**Comment by danielfaust** [comment]
There seems to be a threshold at zoom level 9 and at level 10.

If the zoom level is equal or above 10, then latitudinal and longitudinal clipping occurs.

If the zoom level is equal or above 9 but below 10, then only longitudinal clipping occurs.

If the zoom level is below 9, then no clipping occurs.

So I assume that the threshold equal or above 9 applies the longitudinal clipping and equal or above 10 the latitudinal.

I have no idea where these numbers come from, but they could be a hint.

**Comment by prozessor13** [comment]
Thx for bugreport, i looks like that only the center-coordinate of the image is rendered onto terrain. When looking into *draw_raster* for ImageSource a `boundsBuffer` object is transfered to the shader. But because terrain-tile rendering stops on tile-borders this looks not sufficient. In this case i think a list of terrain-tiles, calculated by the `source.coordinates`, must be rendernd.
Currently i not have the time to look further onto this, anyone else?

**Comment by HarelM** [comment]
I'll see if this is something I can try and tackle, sounds fairly straight forward from your deception.

**Comment by HarelM** [comment]
Unfortunately I'm not making any progress :-(
I've created the following debug page in order to be able to better debug this issue:
Clicking zoom one time shows the issue and allows changing the code using `npm run start-debug` to see the changes in the code take effect on the page.
I looked at the draw_raster `tileIDs` parameter and it is the same between with and without terrain (it's probably related to `boundsBuffer` as you said, but I don't know how to make some thing out of it).
I guess I just am not familiar enough with the code and where to look for the issue... :-(
Leaving it to someone else, if someone is up for it.

```html
<!DOCTYPE html>
<html>

  <head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>3D Map with image source clipped</title>
    <script src="../../dist/maplibre-gl-dev.js"></script>
    <link href="../../dist/maplibre-gl.css" rel="stylesheet" />
    <style>
      body { margin: 0; padding: 0; }
      #map { position: absolute; top: 0; bottom: 0; width: 100%; }
    </style>
  </head>

  <body>
    <div id="map"></div>
    <script>
      const center = [9.3948749, 48.4968956];
      var la_t_tile = center[1] + 0.025;
      var lo_l_tile = center[0] - 0.035;
      var la_b_tile = center[1] - 0.025;
      var lo_r_tile = center[0] + 0.035;
      const map = new maplibregl.Map({
        container: 'map',
        hash: true,
        style: {
            version: 8,
            sources: {
                image: {
                    type: 'image',
                    url: 'https://i.imgur.com/3IyX7CK.png',
                    coordinates: [
                        [lo_l_tile, la_t_tile],
                        [lo_r_tile, la_t_tile],
                        [lo_r_tile, la_b_tile],
                        [lo_l_tile, la_b_tile]
                    ]
                },
                terrain: {
                    "type": "raster-dem",
                    "url": "https://api.maptiler.com/tiles/terrain-rgb/tiles.json?key=get_your_own_OpIi9ZULNHzrESv6T2vL",
                    "tileSize": 256,
                }
            },
            layers: [{
                'id': 'image',
                'type': 'raster',
                'source': 'image',
            }],
            terrain: {
                source: "terrain",
                exaggeration: 1,
            }
        },
        center: center,
        zoom: 11.5,
      });
      
      map.addControl(new maplibregl.NavigationControl({
        visualizePitch: true,
        showZoom: true,
        showCompass: true
      }));

      map.on('load', () => {
        map.showTileBoundaries = true;
      });
    </script>    
  </body>
</html>
```

**Comment by navilaufm** [comment]
Seems I am having the same issue, so I will post my demo links here and close the issued  #1979 I opened.

and the following appears in console:
83209271-e541-4d31-9f74-52a1663491ad:1 Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently

Link to Demonstration
without 3D
https://climaya.com/maplibre/mapl.html?url=https://nube1.on.gt/icon/min_temp_72.png&lon=-98&lat=17&z=4.6&p=28.5&x1=-120.125&x2=-49.875&y1=-0.125&y2=35.125&opacity=0.5

With 3D enabled image overlay does not show complete and opacity is not respected.
https://climaya.com/maplibre/mapl.html?url=https://nube1.on.gt/icon/min_temp_72.png&lon=-98&lat=17&z=4.6&p=28.5&x1=-120.125&x2=-49.875&y1=-0.125&y2=35.125&opacity=0.5&d3=1

**Comment by HarelM** [comment]
Assigned L bounty.
Link to parent Bounty: https://github.com/maplibre/maplibre/issues/189

**Comment by HarelM** [comment]
Sure, please do.

**Comment by manhcuongincusar1** [comment]
Hi @HarelM after investigating, it seems that the partial rendering problem occurs when the image spans multiple tiles but only one tile's terrain data is utilized, leading to incorrect rendering: https://github.com/maplibre/maplibre-gl-js/blob/1cbb496f5cdf6751475986562eb0067c3ace7c4b/src/render/draw_raster.ts#L64
This issue seems quite complicated. Can you consider to extend the bounty of this issue?

(In this screenshot, the photo belongs to tile 11/1088/718. If terrain is enabled, only the partial area inside the red rectangle will be rendered. Only at lower zoom levels (higher elevation) where the image fits into a single terrain tile. At those levels, the image will be rendered fully.)
<img width="808" alt="Screenshot 2023-06-29 at 13 49 57" src="https://github.com/maplibre/maplibre-gl-js/assets/63287450/bd15a4af-9010-4ce0-94b5-13bae845751e">

**Comment by danielfaust** [comment]
I upgraded the demo to v3.1.0 and replaced the defunct Maptiler tiles & terrain with the demotiles from this project

https://jsbin.com/yuhelepevo/edit?html,output

**Comment by manhcuongincusar1** [comment]
@danielfaust I think this problem will happen with all of un-tiled data sources like Image, Heat maps, 3D meshes (which not loaded from a tiled dataset). Mapbox solved this problem by utilizing ProxiedTileId class to map actual tiles to others terrain tiles to solve the problem in source cache level. That is promising idea other way is update function getTerrainData() to return postMatrix from all spanned tiles.

**Comment by manhcuongincusar1** [comment]
Thank you @danielfaust, I will ping you here when there's progress

**Comment by manhcuongincusar1** [comment]
@danielfaust Do have any ideas about the problem? I would love to here from you?

**Comment by danielfaust** [comment]
Here's a modified demo with 2 animated source images.

https://jsbin.com/juwikuq/4/edit?html,output

I have absolutely no idea what the problem is.

**Comment by danielfaust** [comment]
With a bigger radius: https://jsbin.com/juwikuq/7/edit?html,output

It reminds me of another issue regarding of which scale of tiles is shown at which zoom level, apparently there is a mismatch. And since here every second tile is affected, it seems to suggest that the image is living in another zoom level or something like that.

I'm unable to find the issue number, but I think that it was 2d-related as well.

Off Topic, but maybe related: https://github.com/maplibre/maplibre-gl-js/issues/1077 because at the moment when the image center crosses a tile boundary (of the non-visible zoom level), it flickers.

**Comment by manhcuongincusar1** [comment]
Thank for the reference @danielfaust

**Comment by HarelM** [comment]
What bounty did you had in mind?

**Comment by manhcuongincusar1** [comment]
> What bounty did you had in mind?

I suggest XL @HarelM

**Comment by HarelM** [comment]
Done.

**Comment by manhcuongincusar1** [comment]
@HarelM thank you very much

**Comment by danielfaust** [comment]
Added some controls to modify size and other things: https://jsbin.com/bucosel/2/edit?html,output

**Comment by bogdan799** [comment]
Is this somehow helpful? 
![image](https://github.com/maplibre/maplibre-gl-js/assets/17851659/b32a7934-1f3f-4a78-a978-88048d3d00c9)

**Comment by Panthro22** [comment]
When setting an image to a tile level z = 1, why would part of the image applied to the terrain stop rendering when zoomed in to 16-18? I want to understand better some of the constraints applied to raster image source. The image handles just fine until you zoom in near street level. It disappears for portions of tiles that are too close, even though other tiles further from the camera seem to be rendering fine. Once moved closer to those tiles, will it disappear as well? Is there a cutoff that the child stops rendering an image from a parent tile of it or is there a different constraint?

**Comment by HarelM** [comment]
Is this a question about the behavior of image source or this bug? It is unclear from what you wrote. 
If this is not related to this bug please open a different discussion or issue.

**Comment by Panthro22** [comment]
I'm working on a solution to ensure that the image fits within the bounds of the tile the source is assigned to so the image correctly displays without being cut off when you reach the tile level it is assigned to. One of the worst cases is that the image is on the border of a tile and has to be based on the lowest tile level of 0. The solution allows the image to display correctly until you zoom in to tile 15. Then, it stops rendering the part of the image that the tile 15 contains. I have noticed that when I set it to tile level 4, I then can zoom in further before this issue happens. I wonder if I am also running into another issue where the image stops being displayed by a specific difference in tile level. I want to ensure this solution won't contain any other bugs regarding the image not being correctly rendered at all needed tile levels.
Thank you for any guidance that you can give!

**Comment by HarelM** [comment]
This doesn't answer my question. Please open a different issue or discussion and delete the comments from this big report.

**Comment by dennemark** [comment]
Hi,
added a PR with failing test and example for the repo. But I am stuck, too. I do not understand how the `wrap` of `OverscaledTileID` works as described in the PR. Currently the tiles with wrap `-1` and `1` fail in my case.
But actually, moving the coordinate of my cat image example from `11.43085` to `11.45085` returns the same tiles, but the cat is visible.

**Comment by dennemark** [comment]
Ok one more thing:
If I start on this position: [11.43085, 47.32574]
With coordinates at:
```
                const topLat = center[1] + 0.05;
                const leftLng = center[0] - 0.03;
                const bottomLat = center[1] - 0.05;
                const rightLng = center[0] + 0.07;
```
Image is shown as above.
But if I start at: [11.45085, 47.32574]
The cat is shown correct.

So loading of terrain sources is dependent on transform and somehow the raster image does get proper updated source cache of the terrain. I am a bit lost what is happening at what time.

**Comment by dennemark** [comment]
Image source bounds are bigger than its canonical tile. But it seems only the canonical tile is loaded when reading from its source_cache. Therefore only that tile will be rendered.

Image_source has this tile loaded with wrapped tiles ( 1 and -1 wrapped tiles are used on the edges of the map, to continue rendering it left and right side of a world map, i believe). And it stores it on its this.tiles[0/-1/1].
The loadTile function is called from source_cache from here

https://github.com/maplibre/maplibre-gl-js/blob/9b139a8e158224597661f4cf43f9963e62203a99/src/source/source_cache.ts#L857

And it should be called only once, otherwise the url will be loaded multiple times. That's why it would be great to have a proxy to this tile, since we can then acces the tile with a different tileID.
We would need to do this for all tiles that are within the image_source bounds.
I tried to do it hardcoded but did not succeed.
```
 if (!tile) {
            tile = new Tile(tileID, this._source.tileSize * tileID.overscaleFactor());

            this._loadTile(tile, tileID.key, tile.state);
        }

        tile.uses++;
        this._tiles[tileID.key] = tile;
// new code
 if(this._source.type === 'image' && tileID.key !== 'vhvkbb'){
            const clone = tileID.clone()
            clone.canonical.x--
            clone.key = 'vhvkbb' // key of the missing tile with x: 1088 in my example file

            const proxy = new Proxy(tile, {
                get(target, prop, receiver) {
                    if(prop === "tileID"){
                        return clone
                    }
                    return target[prop]
                }
            })
            this._tiles['vhvkbb'] = proxy
        }
```
If I check the coords arriving in draw_raster, the 'vhvkbb' tile id is missing. And the reason seems to be somewhere in `render_to_texture` `prepareForRender` function. The terrain tries to get all renderable tiles, but somehow the 'vhvkbb' tile is missing.

I have to pause my exploration here. Hope someone can pick it up.

**Comment by HarelM** [comment]
@manhcuongincusar1 are you still working on this issue or can I remove your assignment?

**Comment by manhcuongincusar1** [comment]
@HarelM, please remove me from this issue. I don’t have time for it right now. Sorry about that.

**Comment by danielfaust** [comment]
In https://jsbin.com/yutuyes/2/edit?html,output the map is updated to 4.5.0. The image oddly sticks ("bakes in") to some tiles, the entire 4.x-range has this problem.

**Comment by zouyaoji** [comment]
+1

**Comment by pstaszek** [comment]
Hi!
I would like to spend some time and investigate the issue. If it is ok for you, you can give me an assignment.

**Comment by pstaszek** [comment]
Hey!
I've made some progress, I would like to give you an update.

The issue was caused by the way how map is rendered on terrain. Map is split into terrain tiles. Then In `RenderToTexture.prepareForRender()` other source tiles are match to terrain tiles and rendered to texture which is applied on mentioned terrain tiles. Terrain tiles can have different size than other tiles. Image belongs only the one tile. Without terrain, the stencil mask is not applied to image source, which solves the issue when image exceeds one tile. For terrain there is no similar logic. Image should be rendered on all terrain tiles overlapped by the image, but it is not.

I've made small PoC to proof this diagnosis. What I did, I added rendering of all image layers on all terrain tiles. It requires also to modify matrices for those tiles. You can check the code on [this branch](https://github.com/pstaszek/maplibre-gl-js/tree/issue-1559-image-rendered-partially-on-terrain).

https://github.com/user-attachments/assets/592872aa-b329-4725-a1a6-46dc4ed5c047

**Comment by HarelM** [comment]
Nice progress!
My initial idea was to generate a tile pyramid from the image to solve this for everything basically, not sure how intrusive this is compared to what you wrote, looks like what you wrote is simpler...
Does this fix also solve the following issue (and other linked issues that I have marked as duplicate)?
- #4088

**Comment by danielfaust** [comment]
Hi @pstaszek 

Thank you for your effort.

Is this the core change of the solution?

https://github.com/pstaszek/maplibre-gl-js/commit/6d5ffc2fcea33f3ff0f0a6cb14175bc37290dbca#diff-a39025d6bae5c7f234acc580ab732d25a3b4cf3406cf2cda81a5030100f98711R101

**Comment by danielfaust** [comment]
It's not solving the problem for me.

I cloned https://github.com/pstaszek/maplibre-gl-js/tree/issue-1559-image-rendered-partially-on-terrain and ensured that it is really using the modified code by adding a `console.log` before `const imageSourceIds = Object.keys(style.sourceCaches).filter(sourceId => style.sourceCaches[sourceId].getSource().type === 'image');`, and it was printing the log message but still showing the problem.

I am using this test page https://gist.github.com/danielfaust/12fc45d623a2dd4d334cb0ef4d7854e2 via `<address>:9966/test/examples/_image_test.html#11.41/47.278/11.4266`

**Comment by pstaszek** [comment]
Hey! @danielfaust 
Sorry, maybe I should mention that - the PoC I've made isn't the correct solution for the problem, I made it just to validate the hyphothesis. I believe that problem I have described is the major one, but the solution doesn't cover some other problems related to culling or matrices calcuation. I will work on the solution that will cover all cases/zoom/pitch etc.

@HarelM 
Apparently it doesn't, but those issues might be related.

**Comment by pstaszek** [comment]
@danielfaust 
btw I don't think the issue for image at rest was solved in 5.0.0 - I tried this version, also I'm working on the current `main` and for both the issue persists.

**Comment by pstaszek** [comment]
Hi @danielfaust 

I found the reason why your case didn't work. If terrain is registered, there is a callback defined that clears terrain texture cache only for updated tile(s) - for the image it was only the tile where the image belongs to. You can check my branch - I applied dirty fix (update all terrain for image source) https://github.com/pstaszek/maplibre-gl-js/blob/5f0cf09d9253a68d0467a28b3ff9766e605a39d2/src/ui/map.ts#L2090-L2094. I will cover that in final implementation.

**Comment by pstaszek** [comment]
I believe I solved the issue - the solution is on branch: https://github.com/pstaszek/maplibre-gl-js/tree/issue-1559-image-rendered-partially-on-terrain

I now work on tests and code quality.

**Comment by HarelM** [comment]
Thanks! I've briefly looked at the code, is it possible to solve this by creating an images pyramid and load the relevant images much like raster source?

**Comment by pstaszek** [comment]
Sorry for the late response.

Let me start with explaining the solution. Image is rendered always in one tile, but the content could exceed that tile. This approach doesn't work for terrain. Content of each terrain tile is rendered separately, and for each terrain tiles MapLibre renders only overlapping tiles. This is why sometimes image is not rendered - because it is on the tile that does not belong to some terrain tile.

As the solution I added optional information to the `OverscaledTileID` what tiles on various zoom levels are overlapping with this tile. This way while computing tiles to render for each terrain tile, we are able to render image tile as well, even if canonical tile itself are not overlapping (but the content is).

I gave it a thought, but I don't see easier and better solution for now. I can't imagine any solution with creating mulitple tiles for the image that seems better by the design. Terrain tiles could be at any zoom level, also sometimes terrain tiles from various zoom levels are rendered at the same time - so I don't know what image tiles should be created to match them.

Could you elaborate maybe a bit what do you mean by images pyramid? Maybe I miss some ideas.

**Comment by HarelM** [comment]
I think you understood me correctly. 
Image pyramid is taking an image and creating the relevant part in a relevant tile - making a tile pyramid that has the relevant part of the image in the relevant zoom, x, y.
In theory, this process can be made up front or per request. 
If you have an image, a location and zoom, you can calculate which part of the image will appear in which tile and to slice the relevant part of the image - creating a raster like source. 
I tend to think that it will work better in most cases whereas a different solution might fall short in some cases.
But I might be wrong...

**Comment by pstaszek** [comment]
I see your point.

Splitting the image upfront - I'm not sure if it is possible. Consider following example:
We have an image that fits ~single tile on zoom level 6. We have to create 4 another rasters on ZL7, 16 images for ZL8 etc. etc. We don't know on which ZL user lands, so we don't know where to stop.

Doing this on request is certainly doable, however I can't say which approach will be faster:
- rendering the same, whole image for each terrain tile (like now), or
- split the image, and render only this piece.
And I don't see a way to check that, instead of implementing quite complex logic. 

Taking that into account what I would propose is to go with current, simpler approach, and if there will be an issue about the performance, then try the alternative. What do you think?

**Comment by HarelM** [comment]
Yes, I agree creating an image atlas up front is not a feasible direction, as you said.
I'm OK with the current approach.
My main concern is less about performance but more about bugs that might arise from different tile selection and other logic that is somewhat "solved" when taking into consideration raster source.
I've seen all kind of bugs related to globe and tile selection to know that messing with that logic is not as straight forward as it may look.
But in any case, It's probably better than the current bug that exists, so let's proceed with this direction.

**Comment by pstaszek** [comment]
Posting a link to the Open Collective expense: https://opencollective.com/maplibre/expenses/251507
